runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ConstructionService.java

```diff
@@ -0,0 +1,214 @@
+
+  /**
+   * A registrar that creates {@link TransformProvider} instances from {@link
+   * RunnerApi.FunctionSpec}s.
+   *
+   * <p>Transform authors have the ability to provide a registrar by creating a {@link
+   * ServiceLoader} entry and a concrete implementation of this interface.
+   *
+   * <p>It is optional but recommended to use one of the many build time tools such as {@link
+   * AutoService} to generate the necessary META-INF files automatically.
+   */
+  public interface ConstructionServiceRegistrar {
+    Map<String, TransformProvider> knownTransforms();
+  }
+
+  /**
+   * Provides a mapping of {@link RunnerApi.FunctionSpec} to a {@link PTransform}, together with
+   * mappings of its inputs and outputs to maps of PCollections.
+   *
+   * @param <InputT> input {@link PValue} type of the transform
+   * @param <OutputT> output {@link PValue} type of the transform
+   */
+  public interface TransformProvider<InputT extends PValue, OutputT extends PValue> {
+
+    default InputT createInput(Pipeline p, Map<String, PCollection<?>> inputs) {
+      if (inputs.size() == 0) {
+        return (InputT) p.begin();
+      }
+      if (inputs.size() == 1) {
+        return (InputT) Iterables.getOnlyElement(inputs.values());
+      } else {
+        PCollectionTuple inputTuple = PCollectionTuple.empty(p);
+        for (Map.Entry<String, PCollection<?>> entry : inputs.entrySet()) {
+          inputTuple = inputTuple.and(new TupleTag(entry.getKey()), entry.getValue());
+        }
+        return (InputT) inputTuple;
+      }
+    }
+
+    PTransform<InputT, OutputT> getTransform(RunnerApi.FunctionSpec spec);
+
+    default Map<String, PCollection<?>> extractOutputs(OutputT output) {
+      if (output instanceof PDone) {
+        return Collections.emptyMap();
+      } else if (output instanceof PCollection) {
+        return ImmutableMap.of("output", (PCollection<?>) output);
+      } else if (output instanceof PCollectionTuple) {
+        return ((PCollectionTuple) output)
+            .getAll()
+            .entrySet()
+            .stream()
+            .collect(Collectors.toMap(entry -> entry.getKey().toString(), Map.Entry::getValue));
+      } else if (output instanceof PCollectionList<?>) {
+        PCollectionList<?> listOutput = (PCollectionList<?>) output;
+        return IntStream.range(0, listOutput.size())
+            .boxed()
+            .collect(Collectors.toMap(index -> "output_" + index, listOutput::get));
+      } else {
+        throw new UnsupportedOperationException("Unknown output type: " + output.getClass());
+      }
+    }
+
+    default Map<String, PCollection<?>> apply(
+        Pipeline p, String name, RunnerApi.FunctionSpec spec, Map<String, PCollection<?>> inputs) {
+      return extractOutputs(
+          Pipeline.applyTransform(name, createInput(p, inputs), getTransform(spec)));
+    }
+  }
+
+  private Map<String, TransformProvider> registeredTransforms = loadRegisteredTransforms();
```

Comment:
So we maintain a list of pre-registered transforms for expansion ?

Reply:
Well, it's extensible via the ServiceLoader protocol, but yes. Whoever provides the service builds in whatever they want.